qiskit/transpiler/passes/routing/commuting_2q_gate_routing/commuting_2q_gate_grouper.py

```diff
@@ -0,0 +1,120 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""This pass collapses 2-qubit gates that commute into :class:`.Commuting2qBlock`."""
+
+from qiskit.circuit import Qubit, Gate
+from qiskit.dagcircuit import DAGCircuit, DAGOpNode
+from qiskit.transpiler.passes.optimization.commutation_analysis import CommutationAnalysis
+from qiskit.transpiler import TransformationPass, TranspilerError
+
+from qiskit.transpiler.passes.routing.commuting_2q_gate_routing.commuting_2q_block import (
+    Commuting2qBlock,
+)
+
+
+class Commuting2qGateGrouper(TransformationPass):
+    """A pass that collapses 2-qubit gates that commute into :class:`.Commuting2qBlock`."""
+
+    def __init__(self):
+        super().__init__()
+        self._commutation_set_by_node = None
+        self._done_nodes = set()
+        self.requires = [CommutationAnalysis()]
+
+    def run(self, dag: DAGCircuit) -> DAGCircuit:
+        """Collapses 2-qubit gates that commute into :class:`.Commuting2qBlock`.
+
+        Args:
+            dag: A DagCircuit in which commutation analysis was performed.
+
+        Returns:
+            A dag where multi-qubit commuting groups are collapsed as :class:`.Commuting2qBlock`.
+
+        Raises:
+            TranspilerError: If `commutation_set` is not in the property set.
+        """
+        if self.property_set["commutation_set"] is None:
+            raise TranspilerError("commutation_set is not in the property set.")
+
+        new_dag = dag.copy_empty_like()
+
+        for nd in dag.layered_topological_op_nodes():
+            if nd in self._done_nodes:
+                continue
+            if len(nd.qargs) <= 1 or not isinstance(nd.op, Gate):
+                new_dag.apply_operation_back(nd.op, nd.qargs, nd.cargs)
+            else:
+                commuting_op_set = self._commutes_with(nd)
+                used_qarg = {qarg for node in commuting_op_set for qarg in node.qargs}
+                used_carg = {carg for node in commuting_op_set for carg in node.cargs}
+                new_dag.apply_operation_back(
+                    Commuting2qBlock(commuting_op_set),
+                    [qbit for qbit in dag.qubits if qbit in used_qarg],
+                    [cbit for cbit in dag.clbits if cbit in used_carg],
+                )
```

Comment:
this is a good point. A communitation block with a single gate should not be created. Is that what you say?

Reply:
Yeah I think that would address it, basically we should leave gates alone if they're not part of a commuting set since wrapping them just to unwrap them adds a bunch of unnecessary overhead.